type_info.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	typeInfo := TypeInfo{baseTypeInfo: baseTypeInfo{t: TYPE_ENUM}}
+	typeInfo.names = make([]string, len(dict))
+	copy(typeInfo.names, dict)
+	return typeInfo, nil
+}
+
+// ListTypeInfo returns LIST type information.
+func ListTypeInfo(childInfo TypeInfo) (TypeInfo, error) {
+	t := TypeInfo{
+		baseTypeInfo: baseTypeInfo{t: TYPE_LIST},
+		childTypes:   make([]TypeInfo, 1),
+	}
+
+	if childInfo.t == TYPE_INVALID {
+		return t, getError(errAPI, errInvalidChildType)
+	}
+	t.childTypes[0] = childInfo
+	return t, nil
+}
+
+// StructTypeInfo returns STRUCT type information.
+func StructTypeInfo(typeInfos []TypeInfo, names []string) (TypeInfo, error) {
+	t := TypeInfo{
+		baseTypeInfo: baseTypeInfo{
+			t:     TYPE_STRUCT,
+			names: make([]string, len(names)),
+		},
+		childTypes: make([]TypeInfo, len(typeInfos)),
+	}
+
+	if len(typeInfos) == 0 || len(names) == 0 {
+		return TypeInfo{}, getError(errAPI, errEmptySlice)
+	}
+	if len(typeInfos) != len(names) {
+		return t, getError(errAPI, structFieldCountError(len(names), len(typeInfos)))
+	}
+	for i, childType := range typeInfos {
+		if childType.t == TYPE_INVALID {
+			return t, getError(errAPI, addIndexToError(errInvalidChildType, i))
+		}
+	}
+	for i, name := range names {
+		if name == "" {
+			return t, getError(errAPI, addIndexToError(errEmptyName, i))
+		}
+	}
+
+	copy(t.childTypes, typeInfos)
+	copy(t.names, names)
+	return t, nil
+}
+
+// MapTypeInfo returns MAP type information.
+func MapTypeInfo(keyInfo TypeInfo, valueInfo TypeInfo) (TypeInfo, error) {
+	t := TypeInfo{
+		baseTypeInfo: baseTypeInfo{t: TYPE_MAP},
+		childTypes:   make([]TypeInfo, 2),
+	}
+
+	if keyInfo.t == TYPE_INVALID {
+		return t, getError(errAPI, errInvalidKeyType)
+	}
+	if valueInfo.t == TYPE_INVALID {
+		return t, getError(errAPI, errInvalidValueType)
+	}
+
+	t.childTypes[0] = keyInfo
+	t.childTypes[1] = valueInfo
+	return t, nil
+}
+
+func (typeInfo *TypeInfo) logicalType() (C.duckdb_logical_type, error) {
+	switch typeInfo.t {
+	case TYPE_INVALID, TYPE_UHUGEINT, TYPE_ARRAY, TYPE_UNION, TYPE_BIT, TYPE_TIME_TZ:
+		return nil, unsupportedTypeError(unsupportedTypeToStringMap[typeInfo.t])
+
+	case TYPE_BOOLEAN, TYPE_TINYINT, TYPE_SMALLINT, TYPE_INTEGER, TYPE_BIGINT, TYPE_UTINYINT, TYPE_USMALLINT,
+		TYPE_UINTEGER, TYPE_UBIGINT, TYPE_FLOAT, TYPE_DOUBLE, TYPE_TIMESTAMP, TYPE_TIMESTAMP_S, TYPE_TIMESTAMP_MS,
+		TYPE_TIMESTAMP_NS, TYPE_TIMESTAMP_TZ, TYPE_DATE, TYPE_TIME, TYPE_INTERVAL, TYPE_HUGEINT, TYPE_VARCHAR,
+		TYPE_BLOB, TYPE_UUID:
+		return C.duckdb_create_logical_type(C.duckdb_type(typeInfo.t)), nil
+
+	case TYPE_DECIMAL:
+		return C.duckdb_create_decimal_type(C.uint8_t(typeInfo.width), C.uint8_t(typeInfo.scale)), nil
+	case TYPE_ENUM:
+		return typeInfo.logicalEnumType(), nil
+	case TYPE_LIST:
+		return typeInfo.logicalListType(), nil
+	case TYPE_STRUCT:
+		return typeInfo.logicalStructType(), nil
+	case TYPE_MAP:
+		return typeInfo.logicalMapType(), nil
+	}
+	return nil, unsupportedTypeError(unknownTypeErrMsg)
```

Comment:
If we make TypeInfo an interface with `logicalType` as method, this would be handled by the type system. I think it would then also be reasonably to say the `primitiveType` type could then assume the type is valid and supported, and the error could be removed. I think that makes sense, since any issues should have been caught when creating the `TypeInfo` using the methods above, once that is done there should not be any errors.

Reply:
Great suggestion, I've turned `TypeInfo` into an interface.